src/Workspaces/MSBuild/Core/MSBuild/BuildHostProcessManager.cs

```diff
@@ -63,58 +65,98 @@ public async Task<RemoteBuildHost> GetBuildHostWithFallbackAsync(string projectF
-                buildHostProcess.Disconnected += BuildHostProcess_Disconnected;
-
-                // We've subscribed to Disconnected, but if the process crashed before that point we might have not seen it
-                if (process.HasExited)
+                bool reload;
+                do
                 {
-                    buildHostProcess.LogProcessFailure();
-                    throw new Exception($"BuildHost process exited immediately with {process.ExitCode}");
-                }
+                    reload = false;
+
+                    var pipeName = Guid.NewGuid().ToString();
+                    var processStartInfo = CreateBuildHostStartInfo(buildHostKind, pipeName, dotnetPath);
+
+                    var process = Process.Start(processStartInfo);
+                    Contract.ThrowIfNull(process, "Process.Start failed to launch a process.");
+
+                    buildHostProcess = new BuildHostProcess(process, pipeName, _loggerFactory);
+                    buildHostProcess.Disconnected += BuildHostProcess_Disconnected;
+
+                    // We've subscribed to Disconnected, but if the process crashed before that point we might have not seen it
+                    if (process.HasExited)
+                    {
+                        buildHostProcess.LogProcessFailure();
+                        throw new Exception($"BuildHost process exited immediately with {process.ExitCode}");
+                    }
+
+                    // When running on .NET Core, we need to find the right SDK location that can load our project and restart the BuildHost if required.
+                    // When dotnetPath is null, the BuildHost is started with the default dotnet executable, which may not be the right one for the project.
+                    if (buildHostKind == BuildHostProcessKind.NetCore
+                        && projectOrSolutionFilePath is not null
+                        && dotnetPath is null)
+                    {
+                        // The BuildHost will be able to search through all the SDK install locations for a usable MSBuild instance.
+                        var msbuildLocation = await buildHostProcess.BuildHost.FindUsableMSBuildAsync(projectOrSolutionFilePath, cancellationToken).ConfigureAwait(false);
+                        if (msbuildLocation is not null && GetProcessPath() is { } processPath)
+                        {
+                            // The layout of the SDK is such that the dotnet executable is always at the same relative path from the MSBuild location.
+                            dotnetPath = Path.GetFullPath(Path.Combine(msbuildLocation.Path, $"../../{DotnetExecutable}"));
+                            if (dotnetPath is not null && processPath != dotnetPath)
+                            {
+                                // We need to relaunch the .NET BuildHost from a different dotnet instance.
+                                reload = true;
```

Comment:
Is it easier just to call this same method recursively?

Reply:
Sure. But also, how often do you get to write a do while? =)